### ISSUE TITLE
REPL : pureFunctions and captureChecking fail with "language import only allowed at toplevel"

### ISSUE THREAD

**Issue by rjolly** [scala/scala3#16250]
## Compiler version

3.2.2-RC1-bin-SNAPSHOT-git-292e56f

## Minimized code

```scala
scala> import language.experimental.captureChecking
```

## Output

```scala
-- Error: ----------------------------------------------------------------------
1 |import language.experimental.captureChecking
  |                             ^^^^^^^^^^^^^^^
  |                      this language import is only allowed at the toplevel
```

## Expectation

Should work. See https://github.com/lampepfl/dotty/pull/16240/commits/f2c1b24fe02567459285ff33ff11fcf199f931cb

**Comment by rjolly** [comment]
See https://github.com/lampepfl/dotty/pull/16240/commits/f2c1b24fe02567459285ff33ff11fcf199f931cb

**Comment by anatoliykmetyuk** [comment]
This issue was picked for the Issue Spree No. 23 of 08 November 2022 which takes place in a week from now. @rjolly, @mbovel will be working on it. If you have any insight into the issue or guidance on how to fix it, please leave it here.

**Comment by dwijnand** [comment]
Working on imports and the REPL is hard.  I encountered this reality while working on PR #14021 and there's related history that touches on things like withRootImports and the CollectTopLevelImports phase.

**Comment by mbovel** [comment]
#16304 is a work-in-progress PR for this issue, see comments there.

**Comment by rjolly** [comment]
Sorry I wont be able to attend the spree.

**Comment by rjolly** [comment]
This is not right. The intent was instead to fix what the REPL thinks is the toplevel. Using the command line option is not always possible, for instance with the JSR 223 scripting engine.

**Comment by Linyxus** [comment]
I see, I guess we should reopen this issue then?

**Comment by noti0na1** [comment]
The discussion we had was that capture checking import should not be allowed in REPL, to avoid any wired interaction between CCed symbols and non-CC symbols.

**Comment by rjolly** [comment]
Yes, it is still important to restrict it to the toplevel - for any interpreter definition of toplevel, namely one that matches what we get with a command line option. Does that make sense?

**Comment by rjolly** [comment]
I think the problem is that each interpreter input line is embedded in an object before compilation:
```scala
object Line {
  import language.experimental.captureChecking
//                             ^^^^^^^^^^^^^^^
//                    this language import is only allowed at the toplevel
  class Test
}
```
This is so that its terms can be reused later, through an import that is automatically added in the header of subsequent input line/objects. Maybe we could use instead some kind of package or package object? There might be tips in [this discussion](https://contributors.scala-lang.org/t/plan-for-package-objects-in-scala-3/6749).

**Comment by rjolly** [comment]
Specifically, running the repl with -Xprint:typer and a language import that does not fail, we have:
```scala
package <empty> {
  final lazy module val rs$line$1: rs$line$1 = new rs$line$1()
  final module class rs$line$1() extends Object() { this: rs$line$1.type =>
    import language.experimental.genericNumberLiterals
  }
}
```
Maybe we could, for a given input line : gather every imports and move them to outer scope:
```scala
import language.experimental.captureChecking

package <empty> {
  final lazy module val rs$line$1: rs$line$1 = new rs$line$1()
  final module class rs$line$1() extends Object() { this: rs$line$1.type =>
  }
}
```

**Comment by bracevac** [comment]
I'd also like to add my CHF 0.02 here @Linyxus:

The current solution is suboptimal w.r.t. the scaladoc snippet compiler we intend to use in the CC docs. It'd be better to allow the import statement as long as the compiler option was passed. Otherwise, we run into an issue with the language reference, e.g.,
```scala
import language.experimental.captureChecking // <- to make clear we need the import

def foo[C^](x: AnyRef^{C}): AnyRef^{x} = x
```
There it's natural to include (even in invisible "magic comments") the relevant language import for self-contained snippets that can be copy-pasted or even run in Scastie.

However, with the current solution, even if we make sure the snippet compiler is passed
the `-language:experimental.captureChecking` option (and `separationChecking`, or `safe`),
the snippet compiler (which appears to share the REPL logic) will reject that
language import in the snippet.

Towards having a fully checked language reference, we'd be better off if this didn't result in an error as long as the compiler option was passed. I think (?) it would be straightforward if https://github.com/scala/scala3/pull/22220 was refined accordingly.

**Comment by bracevac** [comment]
Capture checking should feel a lot more first class in the REPL now.
